README.md
# Reporebuild

**Reporebuild** is an Shell script, made to be a tool for running local (or not so local) repository with pre-built AUR packages.

## Installation

Just issue this in approriate directory:

    $ git clone https://forge.pztrn.name/pztrn/reporebuild.git

and you're set.

There is no additional dependencies, except of wget (for now) and bash. I'm trying to make this script very sleek and clean.

## Configuration

After first launch reporebuild will create a default configuration file with bad values for you. You should edit `~/.config/reporebuild.conf` by hand and fix them, otherwise I cannot guarantee good work.

## Packages list

List of packages, that will be rebuilt while issuing `reporebuild build all` are placed in `~/.config/reporebuild.list`. Every package should be placed on separate line. E.g.:

    gajim-hg
    psi-plus-git
    ...

## Licensing

For everyday usage **reporebuild** licensed under Beerware license (read BEERWARE for more information). If you making something on top of it - **reporebuild** licensed under Terms and Conditions of GNU General Public License version 2 or higher.


CHANGELOG
[F] - Fixed
[A] - Added
[R] - Removed

Version 1.3
===========

* Completely refactored version.
* Dual-licensed :).


lib/log.lib.sh
#Copyright (C) 2010-2015, Stanislav N. aka pztrn
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License
#as published by the Free Software Foundation; either version 2
#of the License, or (at your option) any later version.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

# This library responsible for logging actions.

# Some things definitions.
# Log string itself
STRING=""

# Debug levels.
declare -A LEVELS
LEVELS["0"]=""
LEVELS["1"]="\033[1;36m[DEBUG]\033[0m "
LEVELS["2"]="\033[1;31m[HARDDEBUG]\033[0m "

# Message types
declare -A TYPES
TYPES["NORMAL"]="\033[1;32m===>\033[0m "
TYPES["WARN"]="\033[1;33m===> WARNING:\033[0m "
TYPES["ERROR"]="\033[1;31m===> ERROR:\033[0m "

# Is that log line is a question?
QUESTION=0

function log() {
    local LEVEL=$1
    shift
    local TYPE=$1
    shift
    local TEXT=$@

    # Make sure that STRING is empty.
    STRING=""

    get_level_format "${LEVEL}"

    # Show arrows only if this isn't a debug message.
    if [[ ${LEVEL} -eq 0 ]]; then
        get_type_format "${TYPE}"
    fi

    STRING+=${TEXT}

    if [[ ${DEBUG} -ge ${LEVEL} ]] && [[ ${LEVEL} -le ${DEBUG} ]]; then
        if [[ $QUESTION -eq 1 ]]; then
            echo -ne ${STRING}
            QUESTION=0
        else
            echo -e ${STRING}
        fi
    fi
}

# Ask user for bool qestions (Yes/No).
function read_bool_answer() {
    local question=$1
    QUESTION=1
    log 0 "WARN" " ${question} "
    QUESTIOIN=0
    read answer
    if [ $answer = "y" -o $answer = "Y" -o $answer = "" ]; then
        return 1
    else
        return 0
    fi
}

function get_level_format() {
    STRING+="${LEVELS[$1]}"
}

function get_type_format() {
    STRING+="${TYPES[$1]}"
}


lib/common.lib.sh
#Copyright (C) 2010-2015, Stanislav N. aka pztrn
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License
#as published by the Free Software Foundation; either version 2
#of the License, or (at your option) any later version.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

# This library responsible for common actions.

# Temporary directory check.
# If temporary directory does not exist - it will create one.
function tmpdir() {
    if [ ! -d /tmp/reporebuild ]; then
        mkdir -p /tmp/reporebuild &> /dev/null
    else
        rm /tmp/reporebuild/{failed.list,reporebuild.tmp} &> /dev/null
    fi
    if [ ! -d $LOGPATH ]; then
        mkdir -p $LOGPATH
    fi
    if [ ! -d $REPOPATH ]; then
        read_bool_answer "Repository path does not exist. Create? "
        if [ $? -eq 1 ]; then
            mkdir -p $REPOPATH
        else
            log 0 "ERROR" "Repository path wasn't created."
            exit 30
        fi
    fi
}

# This function check passed parameters length.
# If it is zero - return 1, otherwise 0.
function check_params_length() {
    params=$@
    if [ ${params[@]} -eq 0 ]; then
        return 1
    else
        return 0
    fi
}


lib/config.lib.sh
#Copyright (C) 2010-2015, Stanislav N. aka pztrn
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License
#as published by the Free Software Foundation; either version 2
#of the License, or (at your option) any later version.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

# This library responsible for any actions with configuration file.

# Configuration check
# If file does not exist - it will create it.
function configcheck() {
    if [ ! -d ~/.config/reporebuild/ ]; then
        mkdir -p ~/.config/reporebuild/
    fi

    if [ -f ~/.config/reporebuild.conf ]; then
        mv ~/.config/reporebuild.conf ~/.config/reporebuild/
    fi
    if [ -f ~/.config/reporebuild.list ]; then
        mv ~/.config/reporebuild.list ~/.config/reporebuild/
    fi

    if [ ! -f ~/.config/reporebuild/reporebuild.conf ]; then
        touch ~/.config/reporebuild/reporebuild.conf
        echo "PKGBUILDSPATH=/data/PKGBUILDS" >> ~/.config/reporebuild/reporebuild.conf
        echo "REPOPATH=/data/WWW/pztrn.ru/repo/i686" >> ~/.config/reporebuild/reporebuild.conf
        echo "REPONAME=pozitpoh" >> ~/.config/reporebuild/reporebuild.conf
        echo "LOGPATH=~/reporebuildlogs" >> ~/.config/reporebuild/reporebuild.conf
        echo "LOG_LINES_COUNT_ON_EXIT=10" >> ~/.config/reporebuild/reporebuild.conf
        echo "# Please, remove next line when configuration will be ok ;)" >> ~/.config/reporebuild/reporebuild.conf
        echo "REMOVETHISLINE=1" >> ~/.config/reporebuild/reporebuild.conf
    fi

    source ~/.config/reporebuild/reporebuild.conf

    if [ "$REMOVETHISLINE" == "1" ]; then
        echo -e "\033[1;31m===> ERROR:\033[0m Please, verify your config located at \033[1;32m~/.config/reporebuild.conf\033[0m"
        exit 255
    fi
}


lib/builder.lib.sh
#Copyright (C) 2010-2015, Stanislav N. aka pztrn
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License
#as published by the Free Software Foundation; either version 2
#of the License, or (at your option) any later version.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

# This library responsible for all package building actions.

# Actual package building.
function build_package() {
    local PKGNAME=$1
    local PKGBUILD_PATH=$2

    if [ ! -d $PKGBUILD_PATH/$PKGNAME ]; then
        log 0 "ERROR" "PKGBUILD for '${PKGNAME}' not found."
        exit 20
    fi

    cd $PKGBUILD_PATH/$PKGNAME
    log 1 "NORMAL" "Executing makepkg..."
    makepkg -f &> $LOGPATH/reporebuild-$PKGNAME.log
    if [ $? != 0 ]; then
        log 0 "ERROR" "Build error. Last 6 lines from log:"
        echo
        cat $LOGPATH/reporebuild-$PKGNAME.log | tail -n 6
        echo
        log 1 "NORMAL" "Adding '{$PKGNAME}' to list of failed packages..."
        echo "  " $PKGNAME >> /tmp/reporebuild/failed.list
    else
        log 0 "NORMAL" "Rebuild complete. Moving package \033[1;31m$PKGNAME\033[0m to repository dir and update repo database... "
        log 1 "NORMAL" "Updating repository..."
        log 2 "NORMAL" "Removing old package..."
        rm $REPOPATH/$PKGNAME*.pkg.tar.* &> /dev/null
        log 2 "NORMAL" "Moving new package into repository..."
        mv $PKGBUILD_PATH/$PKGNAME/$PKGNAME*.pkg.tar.* $REPOPATH &> /dev/null
        log 2 "NORMAL" "Updating repository metadata..."
        cd $REPOPATH && rm $REPONAME.db.tar.gz &> /dev/null
        repo-add $REPONAME.db.tar.gz *.pkg.tar.gz *.pkg.tar.xz &> /dev/null
    fi
}

# Do some preliminary actions (like PKGBUILD fetching), and launch
# build_package().
function build_from_aur() {
    local PKGNAME=$1

    if [ ! -d $PKGBUILDSPATH/$PKGNAME ]; then
        log 1 "WARN" "Directory for '${PKGNAME}' not found, creating"
        mkdir -p $PKGBUILDSPATH/$PKGNAME
    fi

    log 0 "NORMAL" "(Re)Build package: \033[1;31m$PKGNAME\033[0m"
    read_bool_answer "PKGBUILD for package \033[1;31m$PKGNAME\033[0m will be saved in $PKGBUILDSPATH/$PKGNAME directory. Want to proceed [Y/N]?"

    if [ $? -eq 1 ]; then
        fetch_package_from_aur $PKGNAME $PKGBUILDSPATH

        log 0 "NORMAL" "Building package \033[1;31m$PKGNAME\033[0m..."
        build_package $PKGNAME $PKGBUILDSPATH
    else
        fetch_package_from_aur $PKGNAME /tmp/reporebuild/
        tar -xf $PKGNAME.tar.gz -C /tmp/reporebuild/
        cd /tmp/reporebuild/$PKGNAME
        log 0 "NORMAL" "Fetching Complete. Building package..."
        build_package $PKGNAME /tmp/reporebuild
    fi

}

# Some endless loop for PKGBUILD editing. Kinda like yaourt one.
function edit_pkgbuild(){
    while true; do
        read_bool_answer "NORMAL" "Fetching Complete. Would you like to edit PKGBUILD (will use $EDITOR)? [Y/N] "
        if [ $? -eq 1 ]; then
            if [ "$EDITOR" == "" ]; then
                echo -n -e "\033[1;31m===>\033[0m Editor not specified (e.g. exported from \033[1;32m/etc/profile\033[0m). Enter your favorite editor: "
                read editorans
                export EDITOR="$editorans"
            fi
            $EDITOR $PKGBUILDSPATH/$PKGNAME/PKGBUILD
        elif [ $? -eq 0 ]; then
            break
        fi
    done
}


lib/downloader.lib.sh
#Copyright (C) 2010-2015, Stanislav N. aka pztrn
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License
#as published by the Free Software Foundation; either version 2
#of the License, or (at your option) any later version.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

# Downloader library.

# This method downloads PKGBUILD and neccessary file from AUR, within
# one tarball.
function fetch_package_from_aur() {
    local PKGNAME=$1
    local DESTINATION=$2
    SHORT_PKGNAME=$(echo ${PKGNAME} | cut -c 1-2)
    log 1 "NORMAL" "Saving PKGBUILD to '${DESTINATION}/${PKGNAME}'..."
    QUESTION=1
    log 0 "NORMAL" "Fetching PKGBUILD for: \033[1;31m$PKGNAME\033[0m... "
    mkdir -p $DESTINATION/$PKGNAME
    cd $DESTINATION/$PKGNAME && wget https://aur.archlinux.org/packages/$SHORT_PKGNAME/$PKGNAME/$PKGNAME.tar.gz &>/dev/null
    if [ $? -eq 8 ]; then
        log 0 "ERROR" "Invalid package name!"
        rm -rf $DESTINATION/$PKGNAME &>/dev/null
        exit 8
    fi
    if [ $? -eq 7 ]; then
        log 0 "ERROR" "Protocol error."
        echo "  " $PKGNAME >> /tmp/reporebuild/failed.list
        rm -rf $DESTINATION/$PKGNAME &>/dev/null
        exit 7
    fi
    if [ $? -eq 5 ]; then
        log 0 "ERROR" "SSL verification failed. Cannot download PKGBUILD."
        echo "  " $PKGNAME >> /tmp/reporebuild/failed.list
        rm -rf $DESTINATION/$PKGNAME &>/dev/null
        exit 5
    fi
    if [ $? -eq 4 ]; then
        log 0 "ERROR" "Network failure. No connection established?"
        echo "  " $PKGNAME >> /tmp/reporebuild/failed.list
        rm -rf $DESTINATION/$PKGNAME &>/dev/null
        exit 4
    fi
    if [ $? -eq 3 ]; then
        log 0 "ERROR" "I/O Error. Out of space or inodes?"
        rm -rf $DESTINATION/$PKGNAME &>/dev/null
        echo "  " $PKGNAME >> /tmp/reporebuild/failed.list
        exit 3
    fi
    if [ $? -eq 1 ]; then
        log 0 "ERROR" "Really unknown error. Wget returned 1."
        echo "  " $PKGNAME >> /tmp/reporebuild/failed.list
        rm -rf $DESTINATION/$PKGNAME &>/dev/null
        exit 1
    fi
    tar -xf $PKGNAME.tar.gz -C $DESTINATION
    echo -e "\033[1;32mSUCCESS\033[0m"

    if [ -f /tmp/reporebuild/failed.list ]; then
        log 0 "ERROR" "These packages are failed to be fetched:"
        cat /tmp/reporebuild/failed.list
        exit
    fi
}


actions/help.action.sh
#Copyright (C) 2010-2015, Stanislav N. aka pztrn
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License
#as published by the Free Software Foundation; either version 2
#of the License, or (at your option) any later version.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

# Help application. Does nothing more, than showing help.

function help_main() {
    help
}

function help () {
    clear
    echo "=================================================================================="
    echo -e "\033[1;32mLocal repository rebuild script\033[0m version 1.2"
    echo -e "\033[1;33mUsage:\033[0m"
    echo "reporebuild [OPTION] pkgname1 pkgname2 ... pkgnameN"
    echo
    echo "Options:"
    echo "    build all              rebuild all packages specified in "
    echo "                           ~/.config/reporebuild/reporebuild.list"
    echo ""
    echo "    build pkgname          build specified package from AUR (PKGBUILD will be saved"
    echo "                           in PKGBUILDS directory for later usage with -l)"
    echo ""
    echo "    fetch pkgname          fetch PKGBUILD from AUR without building"
    echo ""
    echo "    build local pkgname    rebuild only specific package from local PKGBUILD"
    echo ""
    echo "    config show            shows current reporebuild configuration"
    echo
    echo "You must create ~/.config/reporebuild/reporebuild.conf! See created one for example."
    echo "PKGBUILDs must be placed in <pkgname> folder (for example, for PSI-PLUS package:"
    echo "psi-plus/PKGBUILD)"
    echo
    echo "List of packages in your local repository must be placed in "
    echo "~/.config/reporebuild/reporebuild.list"
    echo "Example contents of ~/.config/reporebuild/reporebuild.list:"
    echo
    echo "packagename1"
    echo "packagename2"
    echo "..."
    echo "packagenameN"
    echo
    echo "=================================================================================="
}


actions/config.action.sh
#Copyright (C) 2010-2015, Stanislav N. aka pztrn
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License
#as published by the Free Software Foundation; either version 2
#of the License, or (at your option) any later version.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

# Configuration actions application.

# Parameters parser.
function config_main() {
    show_config
}

# This method shows reporebuild configuration.
function show_config() {
    echo -e "\033[1;32mPKGBUILDs located at:\033[1;33m" $PKGBUILDSPATH
    echo -e "\033[1;32mRepository path:\033[1;36m" $REPOPATH
    echo -e "\033[1;32mRepository name:\033[1;37m" $REPONAME
    echo -e "\033[1;32mBuild log located at:\033[1;31m" $LOGPATH
    echo
    echo -e "\033[0mIf this information incorrect - fix it in ~/.config/reporebuild/reporebuild.conf"
    exit 0
}


actions/build.action.sh
#Copyright (C) 2010-2015, Stanislav N. aka pztrn
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License
#as published by the Free Software Foundation; either version 2
#of the License, or (at your option) any later version.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


# Package building appllication.

# This function do parameters parsing.
build_main() {
    local TYPE=$1

    log 2 "NORMAL" "Passed parameters: $@"

    if [ "${TYPE}" == "all" ]; then
        all
    elif [ "${TYPE}" == "local" ]; then
        log 0 "NORMAL" "Executing local packages building for:"
        shift
        PACKAGES=$@
        log 0 "NORMAL" "${PACKAGES}"
        localbuildpkg ${PACKAGES}
    else
        check_params_length $@
        if [ $? -eq 1 ]; then
            log 0 "ERROR" "Package name wasn't specified!"
            exit 19
        fi
        log 0 "NORMAL" "Building packages from AUR:"
        log 0 "NORMAL" $@
        buildfromaur $@
    fi
}

# This function checks for packages list existing.
# This packages list should be placed in $HOME/.config/reporebuild.list.
function check_packages_list_file() {
    if [ ! -f "~/.config/reporebuild/reporebuild.list" ]; then
        log 0 "ERROR" "ERROR: packages list not found!"
        log 0 "ERROR" "Either create one by hand, or use 'build pkgnames'!"
        exit 10
    else
        log 1 "NORMAL" "File with list of packages to rebuild exist."
    fi
}

# Starting point to build all packages.
# Essentially, this method just prints out a message and starts build()
# function.
all () {
    # Check for package listing.
    if [ ! -f ~/.config/reporebuild.list ]; then
        log 0 "ERROR" "No packages to rebuild. Add one in '~/.config/reporebuild/reporebuild.list' and retry."
        exit 18
    fi
    log 0 "NORMAL" "Rebuild all packages.\033[0m Output redirected to $LOGPATH/reporebuild-PKGNAME.log"
    buildpkg
}

# Finish point. Printing list of failed packages (if they're present),
# or just saying, that "Job completed".
buildpkg_done () {
    if [ -f /tmp/reporebuild/failed.list ]; then
        echo -e "\033[1;31m===>\033[0m These packages are failed to built:"
        cat /tmp/reporebuild/failed.list
        exit
    else
        log 0 "NORMAL" "Job completed. Check $LOGPATH/reporebuild-PKGNAME.log for details."
        exit
    fi
    exit
}

# This method iterates over reporebuild.list file and feed packages
# to build_package() function from lib/builder.lib.sh.
buildpkg () {
    tmpdir
    check_packages_list_file

    for PKGNAME in $(cat ~/.config/reporebuild/reporebuild.list); do
        log 0 "NORMAL" "Rebuilding package: \033[1;31m$PKGNAME\033[0m"
        build_package $PKGNAME $PKGBUILDSPATH
    done

    buildpkg_done
}

# This method responsible for package building from AUR.
buildfromaur () {
    PACKAGES=$@
    for PKGNAME in ${PACKAGES}; do
        build_from_aur $PKGNAME
    done

    buildpkg_done
}

# This method responsible for package building from local
# PKGBUILD storage.
localbuildpkg () {
    PACKAGES=$@
    for PKGNAME in ${PACKAGES}; do
        log 0 "NORMAL" "Rebuild package: \033[1;31m$PKGNAME\033[0m"
        build_package $PKGNAME $PKGBUILDSPATH
    done

    buildpkg_done
}
